### PR TITLE
ESQL: Speed up reading many nulls

### PR DESCRIPTION
**Pull request by nik9000**
Sometimes we produce output with a *ton* of `null`s and when we do a significant amount of time is spent on tracking the memory usage of each chunk of null colunms. Whole milliseconds in a request that takes dozens of milliseconds. We can avoid all of this by sharing all constant-null vectors produced for a single block.

When running `FROM *` in ESQL on a data-less coordinating node this cuts reading reading pages from 15% of the time to 3.4% of the time, cutting the entire operation from 634ms to 599ms. Note that #105067 is already open and should save about 540ms from the operation. It's unlikely that combining these two will yield a 59ms operation, but we live in hope.

Relates to #103369

### CODE REVIEW

**Review by costin — APPROVED** [review]
LGTM

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/BlockStreamInput.java [hunk 1]
```diff
@@ -21,4 +44,27 @@ public BlockStreamInput(StreamInput delegate, BlockFactory blockFactory) {
...
+
+    /**
+     * Reads the length for a {@link ConstantNullBlock} from the stream and
+     * returns a {@link ConstantNullBlock} with that size. Subsequent calls
+     * to this method that return a block with the same length will return
+     * the same block.
+     */
+    Block readConstantNullBlock() throws IOException {
+        int positions = readVInt();
+        if (lastConstantNullBlock == null) {
```
Comment (dnhatn) [inline x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/BlockStreamInput.java:56]:
```java
if (lastConstantNullBlock == null || lastConstantNullBlock.getPositionCount() != positions || lastConstantNullBlock.tryIncRef() == false) {
    lastConstantNullBlock = blockFactory.newConstantNullBlock(positions);
}
return lastConstantNullBlock;
```

If we don't retain the reference in this stream, and use `tryIncRef()`, then we could avoid managing the life cycle of this input stream. WDYT? However, I am okay with this approach too.

**Review by dnhatn — APPROVED** [review]
I was thinking of using the local breaker; however, it seems unsafe to use it here. So this approach makes sense to me. LGTM, thanks, Nik!

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/BlockStreamInput.java (hunk 1 above)
Comment (nik9000) [reply x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/BlockStreamInput.java:56]: Spent 20 minutes trying to decide if it was ok to have a reference that may be invalid. I hadn't noticed `tryIncRef`, but that does make sense. Let me think some more! I do feel like "this is a ref, let's count it". But it is a funny ref. And tryIncRef would do the right stuff.

Comment (nik9000) [reply x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/BlockStreamInput.java:56]: I think I'll keep it the way I have it. We can flip it later if we feel the need.

Comment (dnhatn) [reply x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/BlockStreamInput.java:56]: I am okay with that too.

**Pull request merged**